Resolves 5 - Entity Component System

smudge202 opened the pull request:
Some basics added for the Entity Component System.

For the time being, I've assumed Components can be assigned to different instances of an entity (so state management is instance based for entities).

However, only one Component of a given type can be assigned (Type based, not instance). I can't really explain why I thought the inconsistency was a good idea, other than _I had a hunch_. Personally, I would rather Components are responsible for their own Aggregation than having the State engine try to handle it.

This also provides some foundations as per #2 . The TFM's are subject to change, but I believe `netcore50` is accessible to `UAP10` (I'll confirm this at some stage), making us compatible with the Universal Windows Platform.

We're on the latest stable DNX dependencies (beta 7) but beta 8 is due for release; we'll cross that bridge when it arrives.

//cc @herecydev @sblackler @danielcirket @mattridgway

^ _let me know if anyone should be added or would prefer to be removed from cc lists on this repo._

Also... win a t-shirt! https://hacktoberfest.digitalocean.com/

smudge202 committed:
Defined Service Extension and Interface

smudge202 committed:
basic component assignment fleshed out

smudge202 committed:
Safe (no exceptions) assignment added

smudge202 committed:
Moved to instance based entities (not entity type)

smudge202 committed:
Corrected test to match description

smudge202 committed:
Basic retrieval added

smudge202 commented on src/Components/ComponentNotAssignedException.cs:
```diff
@@ -0,0 +1,8 @@
+﻿using System;
```
This class is not in use.

sblackler commented on src/Components/project.lock.json:
```diff
@@ -0,0 +1,1118 @@
+{
```
Why has this been included?

sblackler commented on src/Components/ComponentSystem.cs:
```diff
@@ -0,0 +1,9 @@
+{
+	public interface ComponentSystem
+	{
+		void Assign<Entity, Component>(Entity entity, Component component);
```
Thinking out loud, shouldn't always be `TryAssign` to keep the API consistent with `EntityState`?

sblackler commented on tests/Component.Tests/project.lock.json:
```diff
@@ -0,0 +1,896 @@
+{
```
Same as the other lock file. Kill it with fire.

sblackler commented on .gitattributes:
```diff
@@ -0,0 +1,63 @@
+###############################################################################
```
Since we aren't setting anything, do we actually need this file?

smudge202 replied:
It was auto-generated, and as far as I know, doesn't hurt to be here?

sblackler commented on global.json:
```diff
@@ -0,0 +1,3 @@
+{
+  "projects": [ "src" ]
```
curious: Any reason you didn't add `tests`?

sblackler commented on src/Components/ComponentAlreadyAssignedException.cs:
```diff
@@ -0,0 +1,16 @@
+{
+	public class ComponentAlreadyAssignedException : Exception
+	{
+		private Type _entityType;
```
Do these need exposing / made public readonly properties?

herecydev replied:
Yes, there is zero use for them being stored like this as they can't be accessed! Should be public Type {get; private set;}

smudge202 replied:
As mentioned in issue update, this class will be deleted.

sblackler commented on src/Components/EntityState.cs:
```diff
@@ -0,0 +1,8 @@
+	public interface EntityState
+	{
+		bool TryAdd<Component>(Component component);
+		Component Get<Component>();
```
thought: Can you think of a scenario of where you don't know the component at compile time, thus needing an overload that accepts a `Type`?

sblackler replied:
Similarly, shouldn't there be a TryGet for the case where you don't want an exception?

danielcirket replied:
@sblackler Surely it would be better for Get<Component> to just return null if the component didn't exist though? Not sure it would be a good idea to expect an exception over returning null?

sblackler replied:
@danielcirket null vs exception while valid, is not the point i was aiming for. Its more to do with describing the API so that the sematics are clearer for the consumer. So:

    EntityState state;
    if (entityState.TryGet<MyComponent>(out state))
        // do stuff 

vs

    var state = entityState.Get<MyComponent>();
    if(state != null)
         // do stuff

In the latter, if the implementation of `entityState` changes from returning a null to returning an exception then our code is no longer valid because we are binding our code to a known implementation of `EntityState`. However, in the former, the API is guaranteeing the method call will be consistent (as in - returning a bool, setting the out param).

danielcirket replied:
That's fair, and makes sense, I did think that was more towards what you was aiming for but wanted to throw this out there anyway

smudge202 replied:
The intent is for the API not to throw an exception for missing components, though I'll probably leave ArgumentNull's in. I will also change the API to Only Add/Get (without the Try) I think.

sblackler commented on src/Components/StateManagedComponentSystem.cs:
```diff
@@ -0,0 +1,42 @@
+			return TryAssignInternal(entity, component);
+		}
+
+		private bool TryAssignInternal<Entity, Component>(Entity entity, Component component)
```
What's the value in splitting this out vs doing on L#31 and making L#23 call TryAssign?

smudge202 replied:
Doing so would be null checking twice in the case of Assign. Again, subject to change as a result of API changes mentioned in associated issue.

herecydev commented on src/Components/ServiceExtensions.cs:
```diff
@@ -0,0 +1,21 @@
+	{
+		public static IServiceCollection UseComponentSystem(this IServiceCollection services)
+		{
+			services.TryAdd(DefaultServices());
```
Can do this inline. I'm not sure if you would want to make the services public. After all, that's why they are internal.

smudge202 replied:
You need the ability to add those internal services to a composition graph. This is the way, no?

sblackler replied:
I think he's making the point that `DefaultServices` should be private/internal.

herecydev replied:
Yes. Or inline. Choice is up to you :)

smudge202 replied:
Perhaps. This practice was stolen from MS originally:

https://github.com/aspnet/Mvc/blob/27432d94530405f424e9015c8c5311907bd48b3f/src/Microsoft.AspNet.Mvc/MvcServices.cs#L26

However, I think they've changed their mind:

https://github.com/aspnet/Mvc/issues/1634

Trying to work out which pattern is common practice now.

herecydev replied:
The issue I have primarily is that it's public which [Haok mentioned](https://github.com/aspnet/Mvc/issues/1634#issuecomment-65720794). Also by privatizing it you are promoting reuse which totally doesn't make sense here.

herecydev commented on src/Components/StateManagedComponentSystem.cs:
```diff
@@ -0,0 +1,42 @@
+				.GetOrAdd(entity)
+				.TryAdd(component);
+
+		public EntityState StateFor<Entity>(Entity entity)
```
Is StateFor like a getter of sorts? If it doesn't exist you are adding it which might not be intentional by the consumer?

smudge202 commented on src/Components/ServiceExtensions.cs:
```diff
@@ -0,0 +1,21 @@
+{
+	public static class ServiceExtensions
+	{
+		public static IServiceCollection UseComponentSystem(this IServiceCollection services)
```
For all the natter, not one person pointed out the obvious Use/Add error. :)  Will fix this too (Use for application extensions in my opinion and Add for service extensions).

herecydev replied:
Haha, yes Add purely binds services. Use might trigger the process or provide ordering, in terms of aspnet pipeline.

smudge202 committed:
Resolves most issues identitied on PR

smudge202 committed:
removed project.lock.jsons

smudge202 committed:
Changed Component System API get to be consistent with Entity State API get

smudge202 commented:
Latest update addresses (I believe) everything mentioned except the redundant `.gitattributes` which I could care less about. :)

I've changed the API to a more consistent Assign/Get approach (with no exceptions being thrown anywhere).

@herecydev You had a comment about the Get (previously `StateFor`) performing a `GetOrAdd`. I originally started writing out a lengthy description about why I hadn't followed your suggestion, but the more I _discussed it with myself_ in this comment, the more I changed my mind. I think with the c#6 null propagation (elvis) there should be no harm done. For example:

``` c#
// previous API
var singleComponent = _components.Get(entity).Get<Component>();
// ^ had the side effect of adding a state for the entity if it had no components

// new API
var singleComponent = _components.Get(entity)?.Get<Component>();
// ^ requires elvis for the entity state, but no longer has side effects. 
// Awesome trade off in my opinion (if elvis wasn't about I'd be dubious)
```

smudge202 committed:
Updated test names to match current API

smudge202 committed:
Added unassign

smudge202 commented:
Now includes the `Unassign/Remove` methods mentioned in #5 .

herecydev commented:
Not sure If I still agree. I feel `Get` should only be used for retrieving. The unknown Add in the implementation just isn't reflected in the API in my opinion, producing a side effect of adding. My suggestion would be to call the method `GetOrAdd`.

To me the null propagation has no impact here. You're only asserting that the first `Get` could return null, which is fine. That doesn't mitigate the side effect above.

Thoughts?

smudge202 commented:
@herecydev did I get drunk on relentless or did you not check the code? I thought the new implementation was side effect free?

herecydev commented:
@smudge202 totally didn't read!

ghost commented:
:shipit:
